### ISSUE TITLE
CSRF state token does not match one provided

### ISSUE THREAD

**Issue by erosespinola** [Happyr/LinkedIn-API-client#20]
Getting this error when redirecting after successful log in.
Fatal error: Uncaught Exception: CSRF state token does not match one provided. thrown in ~/HappyR/LinkedIn/LinkedIn.php on line 310. 
At the point when the exception is thrown the state variable is actually empty, which causes the exception, but I can't really figure out why is this happening.

**Comment by Nyholm** [comment]
This happens when the session is not started. I've made a fix when using the $_SESSION. I'm looking into the illuminate storage now..

**Comment by Nyholm** [comment]
Look at the latest dev version. Does it solve your issues?

**Comment by Nyholm** [comment]
I've not pushed Version 0.3.0. Please let me know if you experience the same issues with the new version.

**Comment by djuro** [comment]
Thank you, I have been busy lately. Could not test it yet.

**Comment by johnnyjohnny** [comment]
I've been getting the same issue using the latest version and the simple example.  In particular, refreshes were triggering the exception because they used the redirect_uri from linkedin (which includes $_GET['state'] and $_GET['code'].  One workaround was to make sure the page refreshed after authentication with a 'clean' URI (i.e. no $_GET).   Subsequent refreshes then also use this clean URI.  

Maybe the LinkedIn object needs to check the login state before deciding whether to reuse a code - this seems to trigger the csrf check.

**Comment by Nyholm** [comment]
Hi @johnnyjohnny. 
Thank you for following up on this. I'm not sure I understand what you mean. What triggers the refresh of a page?

When the user returns from linkedin.com with a `$_GET['code']` it does some authentication ([see docs](https://github.com/Happyr/LinkedIn-API-client/blob/master/src/Happyr/LinkedIn/LinkedIn.php#L21)). The `user` and the `access token` is stored in session ([here](https://github.com/Happyr/LinkedIn-API-client/blob/master/src/Happyr/LinkedIn/LinkedIn.php#L255) and [here](https://github.com/Happyr/LinkedIn-API-client/blob/master/src/Happyr/LinkedIn/LinkedIn.php#L347)). When you refresh the page again you do not bother with the `$_GET['code']` or CSRF because you [fetch the user from the session](https://github.com/Happyr/LinkedIn-API-client/blob/master/src/Happyr/LinkedIn/LinkedIn.php#L247). 

The problem seams to be that we cant store stuff to the IlluminateStorage. Is that your opinion too?

**Comment by johnnyjohnny** [comment]
Hi @Nyholm - sorry, nothing triggers the page refresh, I just got the issue when I (for various reasons) refreshed in the browser after login.  I started checking isset($_SESSION['linkedIn_access_token'] rather than using $linkedIn->isAuthenticated() - this hit problems in the getCode() method, which I guess means it's trying to re-use the code and has already nulled the state.  I'm not familiar with IlluminateStorage, unfortunately - I'm using session storage I think.  

I can recreate this bug by running the simple code, logging in via the link and hitting refresh when the "Welcome" message has been shown.  The issue seems to arise in the getAccessToken method (around line 317).  A check is made agains $this->accessToken and fetchNewAccessToken is called if not found, i.e. get a new token using a one-time code.   I think this should check the storage instead or "$this" should be re-populated from storage as part of the init method - I don't think this happens.

**Comment by Nyholm** [comment]
Okey. I can reproduce the error now. Thank you. 
I've made a fix. But I need some help validating it. See #26

**Comment by johnnyjohnny** [comment]
Sorry, new to github - how do I get the code from #26?   I can cut and paste but I think I'm missing something.

**Comment by Nyholm** [comment]
You can see that the pull request is made from Nyholm:master. You will find that "fork" here: https://github.com/Nyholm/LinkedIn-API-client

That repository has all the changes, just download that one.

**Comment by benjamindemoor** [comment]
I still get the error even after using this fork https://github.com/Nyholm/LinkedIn-API-client.
Is there anything extra I should do? 
Fatal error: Uncaught Exception: CSRF state token does not match one provided. thrown in ~/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php on line 309

**Comment by nvier** [comment]
@Nyholm I tested [this patch (8bab9f5)](https://github.com/Nyholm/LinkedIn-API-client/commit/8bab9f529412faefba676a7b36c8797e148a9009) and can confirm it fixes the issue. Thank you.

**Comment by Donovan-Ensicaen** [comment]
I have this same issue "CSRF state token does not match one provided." with the backtrace
```
Type: Happyr\LinkedIn\Exceptions\LinkedInApiException

Message: CSRF state token does not match one provided.

Filename: /home/draafterjv/www/application/vendor/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php

Line Number: 302

Backtrace:

File: /home/draafterjv/www/application/vendor/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php
Line: 341
Function: getCode

File: /home/draafterjv/www/application/vendor/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php
Line: 322
Function: fetchNewAccessToken

File: /home/draafterjv/www/application/vendor/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php
Line: 244
Function: getAccessToken

File: /home/draafterjv/www/application/vendor/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php
Line: 224
Function: getUserFromAvailableData

File: /home/draafterjv/www/application/vendor/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php
Line: 472
Function: getUser

File: /home/draafterjv/www/application/vendor/happyr/linkedin-api-client/src/Happyr/LinkedIn/LinkedIn.php
Line: 129
Function: getUserId

File: /home/draafterjv/www/application/controllers/C_utilisateur.php
Line: 461
Function: isAuthenticated
```

I create the url in my view : 
```
$linkedIn = new Happyr\LinkedIn\LinkedIn('xxxxxxxxxxxxxxxx', 'xxxxxxxxxxxxxxxx');
$url = $linkedIn->getLoginUrl(array('redirect_uri' => 'http://www.mywebsite.com/index.php/C_utilisateur/storeLinkedin'));
```
and get the callback in my method storeLinkedin : 
```
    $linkedIn = new Happyr\LinkedIn\LinkedIn('xxxxxxxxxxxxxxxx', 'xxxxxxxxxxxxxxxx');
    if ($linkedIn->isAuthenticated()) {
        //we know that the user is authenticated now. Start query the API
        $user=$linkedIn->api('v1/people/~:(first-name,last-name,headline,location:(name),skills:(skill:(name)))');
    }	    
```

The storage is empty. What's my error? I tried the version 0.3.1 but it's the same issue :/

**Comment by Nyholm** [comment]
Thank you @nvier for testing. I believe we got more than one issue here.. I've merged #26 and created #29. I believe @Donovan-Ensicaen will have an easier time finding the error with the new #29.
